### PR TITLE
feat(rbac): support custom project role management

### PR DESCRIPTION
**Pull request by hazemadelkhalel**
## What does this PR do?

Fixes # (issue)

### CODE REVIEW

File: packages/server/api/src/app/ee/rbac/rbac.controller.ts [hunk 1]
```diff
@@ -1,15 +1,23 @@
...
+    app.get('/', {}, async (req) => {
         const platformId = req.principal.platform.id
         assertNotNullOrUndefined(platformId, 'platformId')
 
-        return rbacService.list(platformId)
+        return await rbacService.list(platformId)
+    })
+
+    app.post('/get-role', GetRbacRequest, async (req) => {
+        return await rbacService.get(req.body.id, req.body.type)
```
Comment (abuaboud) [inline packages/server/api/src/app/ee/rbac/rbac.controller.ts]: /:id

File: packages/server/api/src/app/ee/rbac/rbac.entity.ts [hunk 2]
```diff
@@ -22,6 +22,10 @@ export const RbacEntity = new EntitySchema<RbacSchema>({
             nullable: false,
         },
         platformId: {
+            type: String,
```
Comment (abuaboud) [inline packages/server/api/src/app/ee/rbac/rbac.entity.ts]: rename the entity to project_role

File: packages/server/api/src/app/ee/rbac/rbac.service.ts [hunk 3]
```diff
@@ -1,16 +1,38 @@
...
-    async list(platformId: PlatformId): Promise<Rbac[]> {
-        const rbacs = await rbacRepo().findBy({ platformId })
+    async get(id: ApId, type?: RoleType): Promise<Rbac | null> {
+        return rbacRepo().findOneBy({ id, type })
+    },
+
+    async getDefaultRoleByName(name: ProjectMemberRole): Promise<Rbac> {
+        const role = await rbacRepo().findOneByOrFail({ name, type: RoleType.DEFAULT })
+        return role
+    },
```
Comment (abuaboud) [inline packages/server/api/src/app/ee/rbac/rbac.service.ts]:
```suggestion
    async getDefaultRoleByName(name: ProjectMemberRole): Promise<Rbac> {
        const role = await rbacRepo().findOneByOrFail({ name, type: RoleType.DEFAULT })
        return role
    },
```
We can delete it

File: packages/server/api/src/app/ee/rbac/rbac.service.ts [hunk 4]
```diff
@@ -19,6 +41,40 @@ export const rbacService = {
         return rbacRepo().save(rbac)
     },
 
+    async createDefaultRbac(platformId: PlatformId): Promise<void> {
+        await rbacRepo().save(rbacRepo().create({
```
Comment (abuaboud) [inline packages/server/api/src/app/ee/rbac/rbac.service.ts]: This is can be a for loop

File: packages/react-ui/src/features/team/lib/user-invitation.ts [hunk 5]
```diff
@@ -23,4 +26,9 @@ export const userInvitationApi = {
       invitationToken: token,
     });
   },
+  getDefaultRole(role: ProjectMemberRole): Promise<Rbac> {
```
Comment (abuaboud) [inline packages/react-ui/src/features/team/lib/user-invitation.ts]: Remove

File: packages/react-ui/src/hooks/authorization-hooks.ts [hunk 6]
```diff
@@ -2,22 +2,26 @@ import React from 'react';
...
-      return rolePermissions[role].includes(permission);
+      return role.permissions.includes(permission);
     },
-    [role],
+    [roleId],
   );
 
-  return { checkAccess, role };
+  return { checkAccess, roleId };
 };
```
Comment (abuaboud) [inline packages/react-ui/src/hooks/authorization-hooks.ts]:
```suggestion

export const useAuthorization = () => {
  const roleId = authenticationSession.getUserProjectRoleId();
  const checkAccess = (permission: Permission) => {
    const { data: role, error } = useQuery({
      queryKey: ['rbac', roleId],
      queryFn: async () => {
        if (!roleId) return null;
        const response = await api.post<Rbac>(`/v1/rbac/get-role`, {
          id: roleId,
        });
        return response;
      }
    });

    const hasAccess = React.useMemo(() => {
      if (!roleId) return true;
      if (error || !role) return false;
      return role.permissions.includes(permission);
    }, [role, error, permission]);

    return { hasAccess, error };
  };

  return { checkAccess, roleId };
};

```

File: packages/shared/src/lib/project/project.ts [hunk 7]
```diff
@@ -32,7 +31,16 @@ export const ProjectUsage = Type.Object({
...
+    projectRole: Type.Union([Type.Object({
+        id: Type.String(),
+        created: Type.String(),
+        updated: Type.String(),
+        name: Type.String(),
+        permissions: Type.Array(Type.String()),
+        platformId: Type.Optional(Type.String()),
+        type: Type.String(),
+        userCount: Type.Optional(Type.Number()),
+    }), Type.Null()]),
```
Comment (abuaboud) [inline packages/shared/src/lib/project/project.ts]: Why not refer ProjectRole there?

File: packages/shared/src/lib/invitations/index.ts [hunk 8]
```diff
@@ -37,7 +45,16 @@ export const SendUserInvitationRequest = Type.Union([
...
+            id: Type.String(),
+            created: Type.String(),
+            updated: Type.String(),
+            name: Type.String(),
+            permissions: Type.Array(Type.String()),
+            platformId: Type.Optional(Type.String()),
+            type: Type.String(),
+            userCount: Type.Optional(Type.Number()),
+        }),
     }),
```
Comment (abuaboud) [inline packages/shared/src/lib/invitations/index.ts:42]: You need to send the id not the whole project role Id.

It makes things harder for the api users (frontend), It also can be security issue If you trust the frontend, but I know we have relations but generally it's anti pattern.

File: docs/embedding/provision-users.mdx [hunk 9]
```diff
@@ -61,7 +61,7 @@ To achieve this, the backend will generate a signed token that contains all the
     | firstName         | First name of the user                           |
     | lastName          | Last name of the user                            |
     | email             | Email address of the user                        |
-    | role              | Role of the user in the Activepieces project (e.g., **EDITOR**, **VIEWER**)           |
```
Comment (jasonfill) [inline docs/embedding/provision-users.mdx:64]: @abuaboud This is not clear in the documentation. Is this to be the projectRoleId now or the name of the role, or are there just 3 options that are all upper cased now? If so now do those map to the project role table?

**Pull request merged**